node/worker/worker.go

```diff
@@ -41,29 +42,80 @@ type Worker struct {
 	shardID uint32
 }
 
+// Returns a tuple where the first value is the txs sender account address,
+// the second is the throttling result enum for the transaction of interest.
+// Throttling happens based on the amount, frequency, etc.
+func (w *Worker) throttleTxs(selected types.Transactions, txsThrottleConfig *shardingconfig.TxsThrottleConfig, txnCnts map[common.Address]uint64, tx *types.Transaction) (common.Address, shardingconfig.TxThrottleFlag) {
+	chainID := tx.ChainID()
+	// Depending on the presence of the chain ID, sign with EIP155 or homestead
+	var s types.Signer
+	if chainID != nil {
+		s = types.NewEIP155Signer(chainID)
+	} else {
+		s = types.HomesteadSigner{}
+	}
+
+	var sender common.Address
+	msg, err := tx.AsMessage(s)
+	if err != nil {
+		utils.Logger().Error().Msg("Error when parsing tx into message")
+	} else {
+		sender = msg.From()
+	}
+
+	// already selected max num txs
+	if len(selected) > (*txsThrottleConfig).MaxTxsPerBlockLimit {
+		return sender, shardingconfig.Unselect
+	}
+
+	// throttle a single sender sending too many transactions in one block
+	if ((*txsThrottleConfig).MaxTxAmountLimit).Cmp(tx.Value()) < 0 {
+		return sender, shardingconfig.Invalid
+	}
+
+	// throttle too large transaction
+	if txnCnts[sender] >= (*txsThrottleConfig).MaxTxsPerAccountInBlockLimit {
+		return sender, shardingconfig.Unselect
+	}
+
+	return sender, shardingconfig.Select
+}
+
 // SelectTransactionsForNewBlock selects transactions for new block.
-func (w *Worker) SelectTransactionsForNewBlock(txs types.Transactions, maxNumTxs int, coinbase common.Address) (types.Transactions, types.Transactions, types.Transactions) {
+func (w *Worker) SelectTransactionsForNewBlock(txs types.Transactions, txsThrottleConfig *shardingconfig.TxsThrottleConfig, coinbase common.Address) (types.Transactions, types.Transactions, types.Transactions) {
 	if w.current.gasPool == nil {
 		w.current.gasPool = new(core.GasPool).AddGas(w.current.header.GasLimit)
 	}
+
+	// used for per account transaction frequency throttling
+	txnCnts := make(map[common.Address]uint64)
+
 	selected := types.Transactions{}
 	unselected := types.Transactions{}
 	invalid := types.Transactions{}
 	for _, tx := range txs {
 		if tx.ShardID() != w.shardID {
 			invalid = append(invalid, tx)
 		}
-		snap := w.current.state.Snapshot()
-		_, err := w.commitTransaction(tx, coinbase)
-		if len(selected) > maxNumTxs {
+
+		sender, flag := w.throttleTxs(selected, txsThrottleConfig, txnCnts, tx)
```

Comment:
i would propose to have throttling logic at addPendingTransaction rather than here. See 1/ in https://github.com/harmony-one/harmony/issues/1267

also let's create a sepereate func for throttling

Reply:
I compute num txs per account so that I can apply the throttle logic just 1 time per block.
Also, I think it's better to keep the throttling logic in 1 place for now at least as with txPool integration, throttling will happen through the txpool filter module.